packages/flutter/lib/src/foundation/debug.dart

```diff
@@ -29,6 +31,51 @@ bool debugAssertAllFoundationVarsUnset(String reason, { DebugPrintCallback debug
   return true;
 }
 
+/// Boolean value indicating whether [debugInstrumentAction] will instrument
+/// actions in debug builds.
+bool debugInstrumentationEnabled = false;
+
+/// Runs the specified [action], timing how long the action takes in debug
+/// builds when [debugInstrumentationEnabled] is true.
+///
+/// The instrumentation will be printed to the logs using [debugPrint]. In
+/// non-debug builds, or when [debugInstrumentationEnabled] is false, this will
+/// run [action] without any instrumentation.
+///
+/// Returns the result of running [action], wrapped in a `Future` if the action
+/// was synchronous.
+Future<T> debugInstrumentAction<T>(String description, FutureOr<T> action()) {
+  if (!debugInstrumentationEnabled)
+    return new Future<T>.value(action());
+
+  Stopwatch stopwatch;
+  assert(() {
```

Comment:
these asserts aren't really saving you much. You're still doing a try/finally block, still comparing the results' types, still creating futures, etc. The overhead saved by not doing the stopwatch part is _de minimis_.

If we want to ignore the flag in release builds, use a pattern like:

```dart
  bool instrument = false;
  assert(() { instrument = debugInstrumentationEnabled; return true; });
  if (instrument) {
    // ...
  } else {
    // ...
  }
```

That way the instrumented code will get optimised out entirely in release builds.